bpf: Don't encrypt on path `hostns -> remote pod`

pchaigno opened the pull request:
In pod-to-pod encryption with IPsec and tunneling, Cilium currently encrypts traffic on the path `hostns -> remote pod` even though traffic is in plain-text on the path `remote pod -> hostns`. When using native routing, neither of those paths is encrypted because traffic from the hostns doesn't go through the bpf_host BPF program.

Cilium's Transparent Encryption with IPsec aims at encrypting pod-to-pod traffic. It is therefore unclear why we are encrypting traffic from the hostns. The simple fact that only one direction of the connection is encrypted begs the question of its usefulness. It's possible that this traffic was encrypted by mistake: some of this logic is necessary for node-to-node encryption with IPsec (not supported anymore) and pod-to-pod encryption may have been somewhat simplified to encrypt `*-to-pod` traffic.

Encrypting traffic from the hostns nevertheless creates several issues. First, this situation creates a path asymmetry between the forward and reply paths of `hostns<>remote pod` connections. Path asymmetry issues are well known to be a source of bugs, from of `--ctstate INVALID -j DROP` iptables rules to NAT issues.

Second, Gray recently uncovered a separate bug which, when combined with this encryption from hostns, can prevent Cilium from starting. That separate bug is still being investigated but it seems to cause the reload of bpf_host to depend on Cilium connecting to etcd in a clustermesh context. If this etcd is a remote pod, Cilium connects to it on hostns -> remote pod path. The bpf_host program being unloaded[1], it fails. We end up in a cyclic dependency: bpf_host requires connectivity to etcd, connectivity to etcd requires bpf_host.

This pull request therefore removes encryption with IPsec for the path `hostns -> remote pod` when using tunneling (already unencrypted when using native routing).

1 - More specifically, in Gray's case, the bpf_host program is already loaded, but it needs to be reloaded because the IPsec XFRM  config changed. Without this reload, encryption fails.

```release-note
Fix path asymmetry when using pod-to-pod encryption with IPsec and tunnel mode.
```

pchaigno committed:
bpf: Don't encrypt on path hostns -> remote pod

In pod-to-pod encryption with IPsec and tunneling, Cilium currently
encrypts traffic on the path hostns -> remote pod even though traffic
is in plain-text on the path remote pod -> hostns. When using native
routing, neither of those paths is encrypted because traffic from the
hostns doesn't go through the bpf_host BPF program.

Cilium's Transparent Encryption with IPsec aims at encrypting pod-to-pod
traffic. It is therefore unclear why we are encrypting traffic from the
hostns. The simple fact that only one direction of the connection is
encrypted begs the question of its usefulness. It's possible that this
traffic was encrypted by mistake: some of this logic is necessary for
node-to-node encryption with IPsec (not supported anymore) and
pod-to-pod encryption may have been somewhat simplified to encrypt
*-to-pod traffic.

Encrypting traffic from the hostns nevertheless creates several issues.
First, this situation creates a path asymmetry between the forward and
reply paths of hostns<>remote pod connections. Path asymmetry issues are
well known to be a source of bugs, from of '--ctstate INVALID -j DROP'
iptables rules to NAT issues.

Second, Gray recently uncovered a separate bug which, when combined with
this encryption from hostns, can prevent Cilium from starting. That
separate bug is still being investigated but it seems to cause the
reload of bpf_host to depend on Cilium connecting to etcd in a
clustermesh context. If this etcd is a remote pod, Cilium connects to it
on hostns -> remote pod path. The bpf_host program being unloaded[1], it
fails. We end up in a cyclic dependency: bpf_host requires connectivity
to etcd, connectivity to etcd requires bpf_host.

This commit therefore removes encryption with IPsec for the path hostns
-> remote pod when using tunneling (already unencrypted when using
native routing).

1 - More specifically, in Gray's case, the bpf_host program is already
    loaded, but it needs to be reloaded because the IPsec XFRM config
    changed. Without this reload, encryption fails.
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -344,13 +344,9 @@ handle_ipv6_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		return encap_and_redirect_with_nodeid(ctx, info->tunnel_endpoint,
-						      info->key, info->node_id,
-						      secctx, info->sec_identity,
+						      info->node_id, secctx,
+						      info->sec_identity,
```
Ack on this part.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -373,14 +369,6 @@ handle_ipv6_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		return DROP_UNROUTABLE;
 	}
 
-#ifdef ENABLE_IPSEC
-	if (info->key && info->tunnel_endpoint) {
-		__u8 key = get_min_encrypt_key(info->key);
-
-		set_encrypt_key_meta(ctx, key, info->node_id);
-		set_identity_meta(ctx, secctx);
-	}
-#endif
```
Could you provide a bit more context on this - it's not constrained to `TUNNEL_MODE`, so what traffic did we even expect here? :thinking:

pchaigno replied:
Hm :thinking: 

This code is only executed if going through cilium_host and coming from the host (see `!from_host` check above). AFAIK, for remote destinations, we only go through cilium_host if the destination is part of a remote pod CIDR and we are running in tunneling mode. In native routing mode, we go straight to the native device.

Example routing table for tunneling (10.0.0.0/24 is the remote pod CIDR):
```
10.0.0.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61 mtu 1373 <- we follow this
10.0.1.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61
10.0.1.61 dev cilium_host scope link 
192.168.56.0/24 dev enp0s8 proto kernel scope link src 192.168.56.11**
```
Example routing table for native routing:
```
10.0.0.0/24 via 192.168.56.12 dev enp0s8 <- we follow this
10.0.1.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61 
10.0.1.61 dev cilium_host scope link 
192.168.56.0/24 dev enp0s8 proto kernel scope link src 192.168.56.11
```

So, as far as I can see, this code is only used for tunneling with IPsec. But obviously, tunneling mode should already be handled by the other bit of code I removed just above.

So why was this added? It was added in commit https://github.com/cilium/cilium/commit/b76e6eb59ffcb6af3cd17275762d859450f9a7aa and that commit explicitly says it is to support "direct routing modes". I found that very suspicious because, per the above, in native routing mode, traffic from the hostns shouldn't even go through cilium_host.

I thus tested it out. I've checked IPsec with native routing mode, with and without endpoint routes. I can confirm that, in all cases, traffic from the hostns is not encrypted when going to a remote pod.

AFAICT, this code is dead. I'm unsure when it died.

julianwiedmann replied:
:amaze: thanks for digging deeper!

If you need to push once more - could you split this change (with its related cleanups) into a separate patch? I don't mind cleaning up dead & confusing code as part of the PR, but it would be good to emphasize that it's not contributing to the actual bugfix for Gray's issue. And makes a potential revert down the road a lot easier.

pchaigno replied:
Yeah, that's a good point. I was thinking yesterday that the explanation was too long to include in the commit description, and apparently was too tired to recognize that as an indication the commit should be split.

julianwiedmann commented on bpf/lib/encap.h:
```diff
@@ -219,15 +214,6 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
 	if (!tunnel)
 		return DROP_NO_TUNNEL_ENDPOINT;
 
-#ifdef ENABLE_IPSEC
-	if (tunnel->key) {
-		__u8 key = get_min_encrypt_key(tunnel->key);
-
-		return encap_and_redirect_nomark_ipsec(ctx, key,
-						       tunnel->node_id,
-						       seclabel);
-	}
-#endif
```
Ack.

Note that as we removed both callers in this patch, this should allow us to also delete `encap_and_redirect_nomark_ipsec()`. And `set_encrypt_key_meta()` as well I believe.

julianwiedmann reviewed (commented):
One question to clarify, mostly to understand if this affects more than just the `hostns -> remote pod` path.

pchaigno committed:
node: Don't encrypt traffic to CiliumInternalIP

For the similar reasons as in the previous commit, we don't want to
encrypt traffic going from a pod to the CiliumInternalIP. This is
currently the only node IP address type that is associated an encryption
key.

Since we don't encrypt traffic from the hostns to remote pods anymore
(see previous commit), encrypting traffic going to a CiliumInternalIP
(remote node) would result in a path asymmetry: traffic going to the
CiliumInternalIP would be encrypted, whereas reply traffic coming from
the CiliumInternalIP wouldn't.

This commit removes that caseand therefore ensures we never encrypt
traffic going to a node IP address.

Reported-by: Gray Lian <gray.liang@isovalent.com>
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

julianwiedmann reviewed (approved):
I like it - just for the fact that it brings us in line with the behaviour in native-routing :grin:. 

Thank you Paul!

pchaigno committed:
bpf: Remove IPsec dead code in bpf_host

TL;DR. this commit removes a bit of dead code that seems to have been
intended for IPsec in native routing mode but is never actually
executed.

These code paths are only executed if going through cilium_host and
coming from the host (see !from_host check above). For remote
destinations, we only go through cilium_host if the destination is part
of a remote pod CIDR and we are running in tunneling mode. In native
routing mode, we go straight to the native device.

Example routing table for tunneling (10.0.0.0/24 is the remote pod CIDR):

    10.0.0.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61 mtu 1373 <- we follow this
    10.0.1.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61
    10.0.1.61 dev cilium_host scope link
    192.168.56.0/24 dev enp0s8 proto kernel scope link src 192.168.56.11**

Example routing table for native routing:

    10.0.0.0/24 via 192.168.56.12 dev enp0s8 <- we follow this
    10.0.1.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61
    10.0.1.61 dev cilium_host scope link
    192.168.56.0/24 dev enp0s8 proto kernel scope link src 192.168.56.11

Thus, this code path is only used for tunneling with IPsec. However,
IPsec in tunneling mode should already be handled by the
encap_and_redirect_with_nodeid call above in the same functions (see
info->key argument).

So why was this added? It was added in commit b76e6eb59 ("cilium:
ipsec, support direct routing modes") to support "direct routing modes".
I found that very suspicious because, per the above, in native routing
mode, traffic from the hostns shouldn't even go through cilium_host.

I thus tested it out. I've checked IPsec with native routing mode, with
and without endpoint routes. I can confirm that, in all those cases,
traffic from the hostns is not encrypted when going to a remote pod.

Therefore, this code is dead. I'm unsure when it died.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pippolo84 reviewed (approved):
LGTM for agent related changes :rocket:
